source/common/router/router.cc

```diff
@@ -35,11 +35,7 @@ uint32_t getLength(const Buffer::Instance* instance) { return instance ? instanc
 
 void FilterUtility::setUpstreamScheme(Http::HeaderMap& headers,
                                       const Upstream::ClusterInfo& cluster) {
-  if (cluster.sslContext()) {
-    headers.insertScheme().value().setReference(Http::Headers::get().SchemeValues.Https);
-  } else {
-    headers.insertScheme().value().setReference(Http::Headers::get().SchemeValues.Http);
-  }
+  headers.insertScheme().value().setReference(cluster.transportSocketFactory().httpScheme());
```

Comment:
I guess this is where my other comment comes from. This feels a bit dirty to me, but now I'm seeing why you did this. I will try to think about a better way of doing this but if you have any suggestions happy to hear them. :)

Reply:
Could we require the name of the actual transport socket be the scheme to put in the headers?

Edit: oh, probably not, as again thinking of QUIC it'd be named QUIC but want to translate to https.
OK, the best alternate I can come up with is a utility that takes socket scheme and spits out out http scheme, which at least keeps HTTP-specific information out of the factory

Reply:
Yeah +1. Something along those lines sounds good to me.

Reply:
Can you elaborate "a utility that takes socket scheme and spits out out http scheme" a bit more? I don't get what you mean exactly.